### ISSUE TITLE
Precision and Recall 

### ISSUE THREAD

**Issue by Sairam13001** [qianyuzqy/TransVOD_Lite#6]
Hi @qianyuzqy, I am running your model on my custom dataset. Very impressive work. Thank you for making the code public.

Currently, when we run swin_b_eval_single.py, we get the mAP values right. I am trying to plot "Precision vs Recall" curve but I am facing some issues.

Could you let me know how I can do that.

**Comment by qianyuzqy** [comment]
Hi, @Sairam13001,

Thanks for your interest.  I am sorry that I haven't plotted the "Precision vs Recall" curve using this code.

Maybe you could refer to this link for plotting the Precision-Recall curve: https://github.com/daxiong3721/MOTR/blob/f8d28533e12bb14a64a6c20038f3063e05810731/util/motdet_eval.py#L38 

I hope it will be helpful to you.

Best wishes,

**Comment by Sairam13001** [comment]
Thank you very much.

**Comment by Sairam13001** [comment]
Hi @qianyuzqy, 

I have a doubt in the calculation of mAP. 

Deformable DETR gives 100 proposals per an image(if we set num_queries to 100). While training, out of these 100 proposals, we select the ones that matches best with ground truth using Bipartite matching and calculate loss for only these proposals right. 

When we run inference, how is a prediction selected as True Positive or False Positive? Is it from the IoU between Ground truth and the prediction or from the logit score? Please let me know.

**Comment by fisal-ITS** [comment]
Hai, Have you been able to plot precision vs recall? Can you tell us how?

**Comment by Sairam13001** [comment]
Hi @fisal-ITS, 

In datasets/coco_eval.py, in accumulate() method, we can get the precision values by using "coco_eval.eval['precision']". This returns precision for multiple IoU thresholds, recall thresholds, category ids, area ranges and max_dets. This is a 5 dimensional array. You can select which precision you want to plot - for eg: coco_eval.eval['precision'][0, :, 0, 0, 0]. 

I defined recall ranges from 0 to 1 like this - "np.arange(0, 1.01, 0.01)". Once we have the precision and recall values, plotting them is straight forward. 

Please let me know if this makes sense or if you need more details.

**Comment by fisal-ITS** [comment]
Hai @Sairam13001 

I am very thankful for your instructions. But I am still not clear about the information. if you are willing, would you please provide a tutorial to make it clearer?

And can you tell us how to display the mAP value because when I do the evaluation, the information that appears is the average precision and recall only. Maybe there is a way for me to display the mAP directly.

**Comment by Sairam13001** [comment]
Hi @fisal-ITS, could yout let me know how many classes your dataset has?

**Comment by fisal-ITS** [comment]
Hi @Sairam13001 thank you for the reply. So i tried to detect 1 class of object

**Comment by Sairam13001** [comment]
1. If you have 1 class in your dataset, please define num_classes=2 in models/deformable_detr_single.py in build():
2. Are you able to run the training part? 
If yes, while evaluating, in engine_single.py evaluate(), when coco_evaluator.accumulate() is called: `if coco_evaluator is not None: coco_evaluator.accumulate()`,

In datasets/coco_eval.py  - accumulate():

`def accumulate(self):`
`  for coco_eval in self.coco_eval.values():`
`    coco_eval.accumulate()`
`  PR_folder = 'prec_recall_curves/'`
`  os.makedirs(PR_folder, exist_ok = True)`
`  np.save(PR_folder + 'precision.npy', coco_eval.eval['precision'])`
`  prec = np.load(PR_folder + 'precision.npy')`
`  y = prec[0, :, 0, 0, 2]`
`  x = np.arange(0, 1.01, 0.01)`
`  plt.xlabel("RECALL")`
`  plt.ylabel("PRECISION")`
`  plt.plot(x, y)`
`  plt.savefig(PR_folder + 'PR_Curve.png')`

This way, we can plot the Precision-Recall curve. prec[0, :, 0, 0, 2] means that we are taking precision at IoU threhold=0.5, recall thresholds=0 to 1, category ids='num_classes in your dataset', area ranges=all and max_dets_per_image=10. You can modify this as per your requirement.

I hope this helps.

**Comment by fisal-ITS** [comment]
Thank you so much @Sairam13001 . I have successfully plotted precision vs recall thanks to your instructions. But the precision plot result that I have returns full one. Any suggestions to solve this problem? 
![image](https://github.com/qianyuzqy/TransVOD_Lite/assets/90088354/b9ad6adb-5b7a-479f-ab4d-d57e4ce9b727)
